hydra_pywr/runner.py

```diff
@@ -180,6 +186,13 @@ def __init__(self, *args, domain="water", **kwargs):
 
         self.limit_nodes_recording = False
 
+        tmpdir = tempfile.gettempdir()
+        self.results_location = os.getenv("PYWR_RESULTS_LOCATION", tmpdir)
+        self.bucket_name = os.getenv("PYWR_RESULTS_S3_BUCKET", 'pywr-results')
+        self.s3_path = hash_with_key(self.scenario_id, os.getenv('PYWR_SECURITY_HASH', 'untQnPjA07r0S64rWzkHI72a0dY6KipWt70EqOWTB6F3izvRy9Ude8SOXshfq1vw'))
```

Comment:
The string came from a hash generator. On any production environment, this would need to be changed using the PYWR_SECURITY_HASH environment variable. Do you have a better solution?

Reply:
A warning where the default is used would help, but this is visible only when the logs are available. There will be nothing to distinguish which key was used in the json that goes into the recorder resource scenario on L697 so an alternative is something like adding a `"default_s3_path": True` pair to the json where the default was used. This allows paths which are "insecure" because somebody could derive them using the default PYWR_SECURITY_HASH to be distinguished from those which use the secret equivalent passed here via the env var.

Or/and just definitely guarantee that every deployment uses the env var secret in which case the default here is never used (except for in tests?) and represents a kind of "fallback that fills in for missing config", but that too sounds like a circumstance that should generate some warning.

Reply:
On second thought, the actual hash key is kind of irrelevant, in that it diesn't need to be deterministic.

So I propose simply generating a key every time with:

`hashkey = hashlib.sha256().hexdigest()`

Reply:
That's equivalent to `hashlib.sha256(b"").hexdigest()` so gives the same `e3b0...b855` digest every time - effectively the same as hardcoding the key string.

We could just initialise this with (pseudo)random bytes with...
```python
from random import randbytes
hashlib.sha256(randbytes(16)).hexdigest()
```

...for a unique and difficult to predict key.

Agree with the principle: because `{self.s3_path}` is recorded for each value in the `url` key of the json, we don't need to be able to recreate the hash it was derived from.

Reply:
Good spot, thanks. Updated.